File: src/main/java/service/MatrixPathSimplifier.java
package service;

import java.util.ArrayList;
import java.util.List;
import lombok.AllArgsConstructor;
import model.Matrix;
import model.MatrixPath;
import model.MatrixPoint;

@AllArgsConstructor
public class MatrixPathSimplifier {

    private final BresenhamPathCreator bresenhamPathCreator;
    private final MatrixPathValidator matrixPathValidator;

    public MatrixPath simplify(Matrix matrix, MatrixPath matrixPath) {
        List<MatrixPoint> newPath = new ArrayList<>();
        newPath.add(matrixPath.getPoints().get(0));

        while (newPath.get(newPath.size() - 1) != matrixPath.getPoints().get(matrixPath.getPoints().size() - 1)) {
            MatrixPoint currentPoint = newPath.get(newPath.size() - 1);
            for (int i = matrixPath.getPoints().size() - 1; i > 0 ; i--) {

                MatrixPath bresenhamPath = bresenhamPathCreator.createPath(currentPoint, matrixPath.getPoints().get(i));
                if (matrixPathValidator.validateMatrixPath(matrix, bresenhamPath)) {
                    newPath.add(matrixPath.getPoints().get(i));
                    break;
                }
            }
        }

        return new MatrixPath(newPath);
    }

}

File: src/main/java/service/MatrixPathSimplifier.kt
package service

import model.Matrix
import model.MatrixPath
import model.MatrixPoint
import java.util.ArrayList

class MatrixPathSimplifier(
    private val bresenhamPathCreator: BresenhamPathCreator,
    private val matrixPathValidator: MatrixPathValidator
) {

    fun simplify(matrix: Matrix?, matrixPath: MatrixPath): MatrixPath {
        val newPath: MutableList<MatrixPoint> = ArrayList()
        newPath.add(matrixPath.points[0])

        var curIndex = 1
        while (newPath[newPath.size - 1] !== matrixPath.points[matrixPath.points.size - 1]) {
            val currentPoint = newPath[newPath.size - 1]
            var nextPoint: MatrixPoint? = null

            while (curIndex < matrixPath.points.size) {
                val bresenhamPath = bresenhamPathCreator.createPath(currentPoint, matrixPath.points[curIndex])
                if (matrixPathValidator.validateMatrixPath(matrix, bresenhamPath)) {
                    nextPoint = matrixPath.points[curIndex]
                    curIndex++
                } else {
                    break
                }
            }
            if (nextPoint != null) {
                newPath.add(nextPoint)
            } else {
                newPath.add(matrixPath.points[curIndex - 1])
            }
        }
        return MatrixPath(newPath)
    }
}
